File: dart/lib/src/http_flags_repository.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'flag_models.dart';

class HttpFlagsRepository {
  HttpFlagsRepository({
    required String url,
    required String token,
    required http.Client httpClient,
  })  : _url = _trimTrailingSlashes(url),
        _token = token,
        _httpClient = httpClient;

  final String _url;
  final String _token;
  final http.Client _httpClient;

  Future<Map<String, Flag>> fetchFlags() async {
    final response = await _httpClient.get(
      Uri.parse('$_url/api/v1/flags'),
      headers: <String, String>{
        'Authorization': 'Bearer $_token',
      },
    );

    if (response.statusCode != 200) {
      throw StateError(
        'bandeira: unexpected status ${response.statusCode}: ${response.body}',
      );
    }

    final decoded = _decodeBody(response.body);
    final apiResponse = ApiResponse.fromJson(decoded);
    final byName = <String, Flag>{};
    for (final flag in apiResponse.flags) {
      byName[flag.name] = flag;
    }
    return byName;
  }

  Map<String, dynamic> _decodeBody(String body) {
    dynamic decoded;
    try {
      decoded = jsonDecode(body);
    } catch (error) {
      throw StateError('bandeira: failed to decode response: $error');
    }

    if (decoded is Map<String, dynamic>) {
      return decoded;
    }
    if (decoded is Map) {
      return decoded.map(
        (key, value) => MapEntry(key.toString(), value),
      );
    }

    throw StateError('bandeira: failed to decode response: expected object');
  }
}

String _trimTrailingSlashes(String value) => value.replaceFirst(RegExp(r'/+$'), '');

File: elixir/lib/bandeira/http_flags_repository.ex
defmodule Bandeira.HTTPFlagsRepository do
  @moduledoc false

  alias Bandeira.Config
  alias Bandeira.FlagModels

  @spec fetch_flags(Config.t()) ::
          {:ok, %{optional(String.t()) => Bandeira.FlagModels.Flag.t()}} | {:error, String.t()}
  def fetch_flags(%Config{} = config) do
    url = normalized_url(config.url) <> "/api/v1/flags"
    headers = [{"authorization", "Bearer " <> config.token}]

    with {:ok, status, body} <- request(config, url, headers),
         :ok <- ensure_ok(status, body),
         {:ok, decoded} <- decode_json(body) do
      {:ok, FlagModels.parse_response(decoded)}
    else
      {:error, reason} -> {:error, reason}
    end
  end

  defp request(%Config{http_client: fun}, url, headers) when is_function(fun, 2) do
    fun.(url, headers)
  rescue
    error -> {:error, "request failed: #{Exception.message(error)}"}
  end

  defp request(_config, url, headers), do: default_http_get(url, headers)

  @doc false
  @spec default_http_get(String.t(), [{String.t(), String.t()}]) ::
          {:ok, non_neg_integer(), binary()} | {:error, String.t()}
  def default_http_get(url, headers) do
    _ = :inets.start()
    _ = :ssl.start()

    req_headers =
      Enum.map(headers, fn {name, value} ->
        {String.to_charlist(name), String.to_charlist(value)}
      end)

    request = {String.to_charlist(url), req_headers}

    case :httpc.request(:get, request, [timeout: 10_000], body_format: :binary) do
      {:ok, {{_http_version, status, _reason}, _resp_headers, body}} ->
        {:ok, status, IO.iodata_to_binary(body)}

      {:error, reason} ->
        {:error, "request failed: #{inspect(reason)}"}
    end
  end

  defp ensure_ok(200, _body), do: :ok

  defp ensure_ok(status, body) do
    {:error, "unexpected status #{status}: #{body}"}
  end

  defp decode_json(body) do
    case Jason.decode(body) do
      {:ok, data} -> {:ok, data}
      {:error, error} -> {:error, "failed to decode response: #{Exception.message(error)}"}
    end
  end

  defp normalized_url(url) do
    String.replace(to_string(url), ~r{/+$}, "")
  end
end
